python/S3_VPC_ENDPOINT_ENABLED/S3_VPC_ENDPOINT_ENABLED.py

```diff
@@ -0,0 +1,424 @@
+# Copyright 2017-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the License is located at
+#
+#        http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+'''
+Description
+   Check whether VPC Endpoint is enabled for each VPC to access Amazon S3.
+
+Trigger
+   Periodic (because VPC Configuration Item does not record Endpoint status)
+
+Reports on:
+   AWS::EC2::VPC
+
+Rule Parameters:
+   None
+
+Scenarios:
+  Scenario 1:
+  Given: No VPC is present while using the describe_vpcs() API call
+   Then: Return NOT_APPLICABLE
+
+  Scenario 2:
+  Given: At least one VPC is present
+    And: The S3 service is not present in the list of "ServiceName" on DescribeVpcEndpoints API
+   Then: Return NON_COMPLIANT on this VPC
+
+  Scenario 3:
+  Given: At least one VPC is present
+    And: The S3 service is present in the "ServiceName" key on DescribeVpcEndpoints API
+    And: The "VpcEndpointState" key value is not "Available"
+   Then: Return NON_COMPLIANT on this VPC
+
+  Scenario 4:
+  Given: At least one VPC is present
+    And: The S3 service is present in the "ServiceName" key on DescribeVpcEndpoints API
+    And: The "VpcEndpointState" key value is "Available"
+   Then: Return COMPLIANT on this VPC
+'''
+import re
+import json
+import sys
+import datetime
+import boto3
+import botocore
+
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::EC2::VPC'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    vpc_list = []
+    account_id = event['accountId']
+    ec2_client = boto3.client('ec2')
+    regex_pattern = r'com\.amazonaws\.\S*s3'
+    endpoint_response = ec2_client.describe_vpc_endpoints()
+    vpc_response = ec2_client.describe_vpcs()
+
+    try:
+        if not vpc_response['Vpcs']:
+            annotate = 'There are no VPCs in the region'
+            return build_evaluation(account_id, 'NOT_APPLICABLE', event, annotation=annotate)
+        for vpc in vpc_response['Vpcs']:
+            vpc_list.append(vpc['VpcId'])
+        for vpc_id in vpc_list:
+            endpoint_response = ec2_client.describe_vpc_endpoints(Filters=[

```

Comment:
not using the `service-name` filter here since service names are tied to region and it'll end up being a hardcoded value for each region. Instead I'm doing the regex match for S3 service. Let me know if this makes sense

Reply:
Yes, agreed. I re-checked and using a regex here is a better option. I don't think we need pagination here as 1000 items are returned by default which seem to be more than enough per VPC.

Reply:
resolving this comment